snapshotter_opts_unix.go

```diff
@@ -20,17 +20,92 @@
 package containerd
 
 import (
+	"context"
 	"fmt"
 
 	"github.com/containerd/containerd/snapshots"
 )
 
+const (
+	capabRemapIDs = "remap-ids"
+)
+
 // WithRemapperLabels creates the labels used by any supporting snapshotter
 // to shift the filesystem ownership (user namespace mapping) automatically; currently
 // supported by the fuse-overlayfs snapshotter
 func WithRemapperLabels(ctrUID, hostUID, ctrGID, hostGID, length uint32) snapshots.Opt {
 	return snapshots.WithLabels(map[string]string{
-		"containerd.io/snapshot/uidmapping": fmt.Sprintf("%d:%d:%d", ctrUID, hostUID, length),
-		"containerd.io/snapshot/gidmapping": fmt.Sprintf("%d:%d:%d", ctrGID, hostGID, length),
-	})
+		snapshots.LabelSnapshotUIDMapping: fmt.Sprintf("%d:%d:%d", ctrUID, hostUID, length),
+		snapshots.LabelSnapshotGIDMapping: fmt.Sprintf("%d:%d:%d", ctrGID, hostGID, length)})
+}
+
+func resolveSnapshotOptions(ctx context.Context, client *Client, snapshotterName string, snapshotter snapshots.Snapshotter, parent string, opts ...snapshots.Opt) (string, error) {
+	capabs, err := client.GetSnapshotterCapabilities(ctx, snapshotterName)
+	if err != nil {
+		return "", err
+	}
+
+	for _, capab := range capabs {
+		if capab == capabRemapIDs {
+			// Snapshotter supports ID remapping, we don't need to do anything.
+			return parent, nil
+		}
+	}
+
+	var local snapshots.Info
+	for _, opt := range opts {
+		opt(&local)
+	}
+
+	needsRemap := false
+	var uidMap, gidMap string
+
+	if value, ok := local.Labels[snapshots.LabelSnapshotUIDMapping]; ok {
+		needsRemap = true
+		uidMap = value
+	}
+	if value, ok := local.Labels[snapshots.LabelSnapshotGIDMapping]; ok {
+		needsRemap = true
+		gidMap = value
+	}
+
+	if !needsRemap {
+		return parent, nil
+	}
+
+	var ctrUID, hostUID, length uint32
+	_, err = fmt.Sscanf(uidMap, "%d:%d:%d", &ctrUID, &hostUID, &length)
+	if err != nil {
+		return "", fmt.Errorf("uidMap unparsable: %w", err)
+	}
+
+	var ctrGID, hostGID, lengthGID uint32
+	_, err = fmt.Sscanf(gidMap, "%d:%d:%d", &ctrGID, &hostGID, &lengthGID)
+	if err != nil {
+		return "", fmt.Errorf("gidMap unparsable: %w", err)
+	}
+
+	if ctrUID != 0 || ctrGID != 0 {
+		return "", fmt.Errorf("Container UID/GID of 0 only supported currently (%d/%d)", ctrUID, ctrGID)
+	}
+
+	// TODO(dgl): length isn't taken into account for the intermediate snapshot id.
+	usernsID := fmt.Sprintf("%s-%d-%d", parent, hostUID, hostGID)
+	if _, err := snapshotter.Stat(ctx, usernsID); err == nil {
+		return usernsID, nil
+	}
+	mounts, err := snapshotter.Prepare(ctx, usernsID+"-remap", parent)
+	if err != nil {
+		return "", err
+	}
+	// TODO(dgl): length isn't taken into account here yet either.
+	if err := remapRootFS(ctx, mounts, hostUID, hostGID); err != nil {
```

Comment:
There's equivalent functionality already present in lower-level runtimes, which would be impacted by a solution like this one that unidirectionally dictates how to handle the rootfs ownership. For example, [Sysbox](https://github.com/nestybox/sysbox) runtime, in the absence of idmapped-mount support for overlayfs (kernel < 5.19) relies on `shiftfs` and/or `chown-metacopy` to considerably expedite this operation.

Would it be possible to expose a config knob that overrides this id-remap logic to allow users to decide where to run this logic? This knob would be deprecated once that snapshotters support more efficient id-remap approaches (probably after 5.19+).

Reply:
Yes, makes sense; it would essentially be an option to say that snapshotters support the "remap-ids" capability even if they don't declare it (and it's handled elsewhere).

(The only downside I see is things outside containerd cannot see this capability, so e.g. if a snapshotter is deciding whether it reports the capability based on the kernel version or not, something like sysbox may have to infer that too, given this would be a user config though I think that's fine. I'll have a little think about that and add something here soon.)

Reply:
> Yes, makes sense; it would essentially be an option to say that snapshotters support the "remap-ids" capability even if they don't declare it (and it's handled elsewhere).

Yes, i think that would work. The snapshotter wouldn't do anything in this case and would rely on lower-level runtimes for this operation.


> (The only downside I see is things outside containerd cannot see this capability, so e.g. if a snapshotter is deciding whether it reports the capability based on the kernel version or not, something like sysbox may have to infer that too...

Agree, we are already doing something like this.